Related information of diagnostics is not parsed & displayed.

### Neovim version (nvim -v)

NVIM v0.8.0-dev-758-g711ef4eac9

### Language server name/version

clangd 14.0

### Operating system/version

Arch Linux

### Steps to reproduce using "nvim -u minimal_init.lua"

Many servers support relatedInformation; I'm personally struggling with clagnd.

```bash
$ cat <<EOF >repro.cpp
void foo(char);
void foo(double);
void bar() {
    foo(1);
}
EOF
$ nvim -u minimal_init.lua repro.cpp
:lua require('lspconfig').clangd.setup{}
:lua vim.diagnostic.setqflist()
```

### Expected behavior

Clangd shows three diagnostics (error + 2x info), just like standalone compiler:
```
repro.cpp:4:5: error: call to 'foo' is ambiguous
    foo(1);
    ^~~
repro.cpp:1:6: note: candidate function
void foo(char);
     ^
repro.cpp:2:6: note: candidate function
void foo(double);
```

VSCode handles this correctly:
![img](https://user-images.githubusercontent.com/17150781/183085819-919ca79b-ce48-4225-b224-19678647d458.png)

### Actual behavior

Clangd reports one diagnostic with two items in the `relatedInformation` list:
https://gist.github.com/BigRedEye/253629638e1d71f4df667a3e8e05a1d3.

Neovim does not parse relatedInformation and displays just one main diagnostic:
![img](https://user-images.githubusercontent.com/17150781/183086488-b136360c-7e23-4fd8-b4b8-0a01f7418fed.png)


### Log file

_No response_

--- a/runtime/doc/lsp.txt
+++ b/runtime/doc/lsp.txt
@@ -1900,6 +1900,12 @@ workspace_symbol({query}, {opts})             *vim.lsp.buf.workspace_symbol()*
 ==============================================================================
 Lua module: vim.lsp.diagnostic                                *lsp-diagnostic*
 
+This module provides functionality for requesting LSP diagnostics for a
+document/workspace and populating them using |vim.Diagnostic|s.
+`DiagnosticRelatedInformation` is supported: it is included in the window
+shown by |vim.diagnostic.open_float()|.
+
+
 from({diagnostics})                                *vim.lsp.diagnostic.from()*
     Converts the input `vim.Diagnostic`s to LSP diagnostics.
 

--- a/runtime/doc/news.txt
+++ b/runtime/doc/news.txt
@@ -196,6 +196,9 @@ LSP
   receives the resolved config as the second arg: `cmd(dispatchers, config)`.
 • Support for annotated text edits.
 • `:checkhealth vim.lsp` is now available to check which buffers the active LSP features are attached to.
+• LSP `DiagnosticRelatedInformation` is now shown in
+  |vim.diagnostic.open_float()|. It is read from the LSP diagnostic object
+  stored in the `user_data` field.
 
 LUA
 

--- a/runtime/lua/vim/diagnostic.lua
+++ b/runtime/lua/vim/diagnostic.lua
@@ -2354,8 +2354,9 @@ function M.open_float(opts, ...)
     end
     local hiname = floating_highlight_map[diagnostic.severity]
     local message_lines = vim.split(diagnostic.message, '\n')
+    local default_pre = string.rep(' ', #prefix)
     for j = 1, #message_lines do
-      local pre = j == 1 and prefix or string.rep(' ', #prefix)
+      local pre = j == 1 and prefix or default_pre
       local suf = j == #message_lines and suffix or ''
       lines[#lines + 1] = pre .. message_lines[j] .. suf
       highlights[#highlights + 1] = {
@@ -2365,8 +2366,38 @@ function M.open_float(opts, ...)
           hlname = prefix_hl_group,
         },
         suffix = {
-          length = j == #message_lines and #suffix or 0,
+          length = #suf,
           hlname = suffix_hl_group,
+        },
+      }
+    end
+
+    ---@type lsp.DiagnosticRelatedInformation[]
+    local related_info = vim.tbl_get(diagnostic, 'user_data', 'lsp', 'relatedInformation') or {}
+
+    -- Below the diagnostic, show its LSP related information (if any) in the form of file name and
+    -- range, plus description.
+    for _, info in ipairs(related_info) do
+      -- TODO: Somehow allow users to open the location when their cursor is over it?
+      local file_name = vim.fs.basename(vim.uri_to_fname(info.location.uri))
+      local info_suffix = ': ' .. info.message
+      lines[#lines + 1] = string.format(
+        '%s%s:%s:%s%s',
+        default_pre,
+        file_name,
+        info.location.range.start.line,
+        info.location.range.start.character,
+        info_suffix
+      )
+      highlights[#highlights + 1] = {
+        hlname = '@string.special.path',
+        prefix = {
+          length = #default_pre,
+          hlname = prefix_hl_group,
+        },
+        suffix = {
+          length = #info_suffix,
+          hlname = 'NormalFloat',
         },
       }
     end

--- a/runtime/lua/vim/lsp/diagnostic.lua
+++ b/runtime/lua/vim/lsp/diagnostic.lua
@@ -1,3 +1,7 @@
+---@brief This module provides functionality for requesting LSP diagnostics for a document/workspace
+---and populating them using |vim.Diagnostic|s. `DiagnosticRelatedInformation` is supported: it is
+---included in the window shown by |vim.diagnostic.open_float()|.
+
 local lsp = vim.lsp
 local protocol = lsp.protocol
 local ms = protocol.Methods

--- a/runtime/lua/vim/lsp/protocol.lua
+++ b/runtime/lua/vim/lsp/protocol.lua
@@ -349,6 +349,7 @@ function protocol.make_client_capabilities()
           valueSet = get_value_set(constants.DiagnosticTag),
         },
         dataSupport = true,
+        relatedInformation = true,
       },
       inlayHint = {
         dynamicRegistration = true,
@@ -540,6 +541,7 @@ function protocol.make_client_capabilities()
         honorsChangeAnnotations = true,
       },
       publishDiagnostics = {
+        relatedInformation = true,
         tagSupport = {
           valueSet = get_value_set(constants.DiagnosticTag),
         },

